wrappers/android/zxingcpp/src/main/java/com/zxingcpp/ZXingCpp.kt

```diff
@@ -0,0 +1,152 @@
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package com.zxingcpp
+
+import android.graphics.Bitmap
+import android.graphics.ImageFormat
+import android.graphics.Point
+import android.graphics.Rect
+import android.os.Build
+import androidx.camera.core.ImageProxy
+import java.nio.ByteBuffer
+
+public object ZXingCpp {
+	private val supportedYUVFormats: List<Int> =
+		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+			listOf(ImageFormat.YUV_420_888, ImageFormat.YUV_422_888, ImageFormat.YUV_444_888)
+		} else {
+			listOf(ImageFormat.YUV_420_888)
+		}
+
+	init {
+		System.loadLibrary("zxing_android")
+	}
+
+	// Enumerates barcode formats known to this package.
+	// Note that this has to be kept synchronized with native (C++/JNI) side.
+	public enum class Format {
+		NONE, AZTEC, CODABAR, CODE_39, CODE_93, CODE_128, DATA_BAR, DATA_BAR_EXPANDED,
+		DATA_MATRIX, EAN_8, EAN_13, ITF, MAXICODE, PDF_417, QR_CODE, MICRO_QR_CODE, UPC_A, UPC_E
+	}
+
+	public enum class ContentType {
+		TEXT, BINARY, MIXED, GS1, ISO15434, UNKNOWN_ECI
+	}
+
+	public enum class Binarizer {
+		LOCAL_AVERAGE, GLOBAL_HISTOGRAM, FIXED_THRESHOLD, BOOL_CAST
+	}
+
+	public enum class EanAddOnSymbol {
+		IGNORE, READ, REQUIRE
+	}
+
+	public enum class TextMode {
+		PLAIN, ECI, HRI, HEX, ESCAPED
+	}
+
+	public data class DecodeHints(
+		var formats: Set<Format> = setOf(),
+		var tryHarder: Boolean = false,
+		var tryRotate: Boolean = false,
+		var tryInvert: Boolean = false,
+		var tryDownscale: Boolean = false,
+		var isPure: Boolean = false,
+		var tryCode39ExtendedMode: Boolean = false,
+		var validateCode39CheckSum: Boolean = false,
+		var validateITFCheckSum: Boolean = false,
+		var returnCodabarStartEnd: Boolean = false,
+		var returnErrors: Boolean = false,
+		var downscaleFactor: Int = 3,
+		var eanAddOnSymbol: EanAddOnSymbol = EanAddOnSymbol.IGNORE,
+		var binarizer: Binarizer = Binarizer.LOCAL_AVERAGE,
+		var textMode: TextMode = TextMode.HRI,
+		var minLineCount: Int = 2,
+		var maxNumberOfSymbols: Int = 0xff,
+		var downscaleThreshold: Int = 500
+	)
+
+	public data class Position(
+		val topLeft: Point,
+		val topRight: Point,
+		val bottomLeft: Point,
+		val bottomRight: Point,
+		val orientation: Double
+	)
+
+	public data class Result(
+		val format: Format,
+		val bytes: ByteArray?,
+		val text: String?,
+		val time: String?, // for development/debug purposes only
+		val contentType: ContentType,
+		val position: Position?,
+		val orientation: Int,
+		val ecLevel: String?,
+		val symbologyIdentifier: String?,
+		val sequenceSize: Int,
+		val sequenceIndex: Int,
+		val sequenceId: String,
+		val readerInit: Boolean,
+		val lineCount: Int
+	)
+
+	public fun read(image: ImageProxy, decodeHints: DecodeHints): List<Result>? {
```

Comment:
Since a `List` can inherently be empty, I wonder whether `List<Restult>` would be the better return type here. I can imagine that the built in syntactic sugar of Kotlin to perform conditional operations based on the value of a nullable/optional would make it more convenient from a client perspective to use the optional as you did here. Just saying that from a conceptual point of view it does make sens to have two different ways of specifying 'nothing found'. If we stick with the optional, we should probably add a comment somewhere that the returned list object is either `null` or non-empty, otherwise you would have to check for nullness and emptiness?

Reply:
> Since a List can inherently be empty, I wonder whether List<Restult> would be the better return type here. 

Hm, yes, that's true! From a client's point of view it probably makes more sense to always return `List<Result>` and cut the nullness case. One thing less to worry about.

The reason why I chose to return `null` (if there are no results) is, that this saves us from allocating an empty `ArrayList` in the hot path. But that's not a big deal, of course, and also probably more of a leaking abstraction. So if you prefer `List<Result>` over `List<Result>?` I will happily change it.

Syntactically it's almost the same to test for nullness or emptiness in Kotlin.

>  If we stick with the optional, we should probably add a comment somewhere that the returned list object is either null or non-empty, otherwise you would have to check for nullness and emptiness?

Agreed, otherwise a comment would make sense 👍 Maybe just before [read()](https://github.com/zxing-cpp/zxing-cpp/pull/659/files#diff-40bfa6be62aab191f820c872aa4229fcaabe9efdc96a6578ee56a2d071d3c175R108)?

Reply:
I believe `List<Result>` is the most natural return type. Thanks.

Reply:
I've changed the return type to `List<Result>` in cf3cb80.